### PR TITLE
Implement custom JVP rule for improving `safe_norm` computation speed

### PR DESCRIPTION
**Pull request by flferretti**
This pull request introduces a custom JVP rule for the `safe_norm` function to handle NaNs and zero arrays more effectively during gradient computation, while having the same speed of the plain `jnp.linalg.norm` when the gradients are not necessary. This enhancement improves the runtime speed of the `safe_norm` function

Note that `@jax.custom_jvp` only supports functions with positional args or positional-or-keyword, for this reason the non-array args (axis, keepdims) should be trated as static. For this reason a wrapper factory was necessary

----
📚 Documentation preview 📚: https://jaxsim--411.org.readthedocs.build//411/

### CODE REVIEW

**Review by younik — COMMENTED** [review]
This is very smart, and the benchmarks look good!

I would maybe add a test for this.

> Note that @jax.custom_jvp only supports functions with positional args or positional-or-keyword, for this reason the non-array args (axis, keepdims) should be trated as static. For this reason a wrapper factory was necessary

I didn't undesrtand this issue. Why do we need the factory wrapper?

**Comment by flferretti** [comment]
> This is very smart, and the benchmarks look good!

Thank you!

> I would maybe add a test for this.

There should be already a autodiff test in https://github.com/ami-iit/jaxsim/blob/b1cabaf93151b7ca9aefd818fcbd67c14b1939ea/tests/test_automatic_differentiation.py#L421-L462

Would you like to add anything to the test?
 
> > Note that @jax.custom_jvp only supports functions with positional args or positional-or-keyword, for this reason the non-array args (axis, keepdims) should be trated as static. For this reason a wrapper factory was necessary
> 
> I didn't undesrtand this issue. Why do we need the factory wrapper?

The reason is `@jax.custom_jvp` only supports positional arguments, and all non-array arguments must be static. In other words, JAX needs to be able to trace the function's behavior deterministically based on the array inputs only.

`axis` and `keepdims` are non-array arguments and they control how the function behaves (e.g., along which axis to compute the norm). If you just pass them normally to a `@custom_jvp` function, JAX would not know how to trace/compile the JVP properly

The wrapper factory like:

```python
@functools.cache
def _make_safe_norm(axis, keepdims):
    @jax.custom_jvp
    def _safe_norm(array):
        ...
    @_safe_norm.defjvp
    def _safe_norm_jvp(...):
        ...
    return _safe_norm
```

solves this by baking in the `axis` and `keepdims` values as **static**. So when you call:

```python
safe_norm(array, axis=1, keepdims=True)
```

under the hood it becomes:

```python
_make_safe_norm(axis=1, keepdims=True)(array)
```

Now JAX only sees a function with a single positional array argument, which is exactly what `@custom_jvp` expects. The `axis` and `keepdims` are static and outside the JAX trace

**Comment by flferretti** [comment]
@ami-iit/darwin friendly ping :)

**Review by CarlottaSartore — APPROVED** [review]

**Review by xela-95 — APPROVED** [review]
Really neat code!! Thanks @flferretti

**Comment by flferretti** [comment]
@ami-iit/darwin I removed the closure caching as I believe it can cause conflicts with JAX and potentially cause cache invalidation of JIT-compiled functions. I tried comparing with and without it with the benchmarks and they are better without the `@functools.cache`
